File: go/proverb/proverb.go
/*
Package proverb provides a method to generate proverb from given list of inputs.
*/
package proverb

import "fmt"

// Proverb returns a proverb from given rhymes.
func Proverb(rhyme []string) []string {
	endingTemplate := "And all for the want of a %s."
	template := "For want of a %s the %s was lost."
	if len(rhyme) == 0 {
		return []string{}
	}

	out := []string{}

	for i, n := 0, len(rhyme)-1; i < n; i++ {
		out = append(out, fmt.Sprintf(template, rhyme[i], rhyme[i+1]))
	}
	out = append(out, fmt.Sprintf(endingTemplate, rhyme[0]))

	return out
}

File: haskell/proverb/src/Proverb.hs
module Proverb(recite) where

proverb :: (String, String) -> String
proverb (p, q) = "For want of a " ++ p ++ " the " ++ q ++ " was lost."

recite :: [String] -> String
recite [] = ""
recite [a] = "And all for the want of a " ++ a ++ "."
recite list = unlines (map (proverb) (zip list $ tail list)) ++ "And all for the want of a " ++ (head list) ++ "."
